File: LeetCode-Swift/LeetCode-Swift/堆/_451_根据字符出现频率排序.swift
//
//  _451_根据字符出现频率排序.swift
//  LeetCode-Swift
//
//  Created by warden on 2020/2/24.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/sort-characters-by-frequency/

import Foundation

class _451_根据字符出现频率排序 {
    func frequencySort(_ s: String) -> String {
        typealias Pair = (Character, Int)
        var m = [Character: Int]()
        for ch in s {
            m[ch] = 1 + (m[ch] ?? 0);
        }
        var pq = [Pair]()
        for item in m {
            pq.append((item.key, item.value))
        }
        pq.sort { (item1, item2) -> Bool in
            return item1.1 > item2.1
        }
        var res = String()
        for item in pq {
            res += String(repeating: item.0, count: item.1)
        }
        return res
    }
}

/**
 示例 1:

 输入:
 "tree"

 输出:
 "eert"

 示例 2:

 输入:
 "cccaaa"

 输出:
 "cccaaa"

 示例 3:

 输入:
 "Aabb"

 输出:
 "bbAa"
 */
func test_451() {
    let solution = _451_根据字符出现频率排序()
    print(solution.frequencySort("tree"))
    print(solution.frequencySort("cccaaa"))
    print(solution.frequencySort("Aabb"))
}

File: LeetCode-C++/LeetCode-C++/堆/_451_根据字符出现频率排序.cpp
//
//  _451_根据字符出现频率排序.cpp
//  LeetCode-C++
//
//  Created by warden on 2020/2/24.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/sort-characters-by-frequency/

#include <stdio.h>
#include <string>
#include <map>
#include <queue>

using namespace std;

class Solution {
public:
    string frequencySort(string s) {
        string res;
        map<char, int> m;
        for(auto val:s)
            m[val]++;

        priority_queue<pair<int, char>> pq;
        for(auto val:m)
            pq.push(make_pair(val.second, val.first));
        
        while (!pq.empty()) {
            auto top = pq.top();
            int num = top.first;
            while(num--)
                res+=top.second;
            pq.pop();
        }
        return res;
    }
};

/**
 示例 1:

 输入:
 "tree"

 输出:
 "eert"

 示例 2:

 输入:
 "cccaaa"

 输出:
 "cccaaa"

 示例 3:

 输入:
 "Aabb"

 输出:
 "bbAa"
 */
int main(int argc, const char * argv[]) {
    Solution solution = Solution();
//    string result = solution.frequencySort("tree");
    string result = solution.frequencySort("cccaaa");
    printf("%s\n", result.c_str());

    return 0;
}
